### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by methodsecuritysj**
Working initial version of `webassess url`

To test the binary locally, ensure you have ollama installed:

```
ollama pull qwen2.5:3b
go run main.go url --target http://basicwebsite-postgresdb.method.directory --output json --ollama-model qwen2.5:3b
```

You can also build things with go releaser and test it running against the docker image

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,35 +1,33 @@
...
 
 # Copy the CLI binary and make it executable
 COPY ${CLI_NAME} /opt/method/${CLI_NAME}/service/bin/${CLI_NAME}
-RUN chmod +x /opt/method/${CLI_NAME}/service/bin/${CLI_NAME}
 
 # Install models, ollama needs to be running both for install and entrypoint
-COPY configs/setup-ollama.sh /opt/method/${CLI_NAME}/var/conf/setup-ollama.sh
+# A single dedicated script is used to have ollama serve be run in the same terminal
+ENV OLLAMA_MODELS=/opt/method/${CLI_NAME}/var/conf/models
+RUN echo '#!/usr/bin/env bash\n\nollama serve &\nsleep 5\nollama list\nollama pull qwen2.5:0.5b\nollama pull qwen2.5:3b\nollama pull gemma2:2b' > /opt/method/${CLI_NAME}/var/conf/setup-ollama.sh
```
Comment (sean-hacker) [inline Dockerfile]: Can we make this a script in the repo instead of an echo?

File: cmd/root.go [hunk 2]
```diff
@@ -28,35 +32,135 @@ type AIAssess struct {
...
+				_, pathErr := exec.LookPath("ollama")
+				if pathErr != nil {
+					errorMessage := "ollama is not installed or is not in the system path"
+					a.OutputSignal.ErrorMessage = &errorMessage
+					a.OutputSignal.Status = 1
+					return fmt.Errorf(errorMessage)
+				}
+
+				// Check to see if ollama is running without being spawned by the CLI
+				if !ollama.IsOllamaRunning("http://127.0.0.1:11434/api/tags") {
```
Comment (sean-hacker) [inline cmd/root.go]: Can we pull this URL out into a constant?

File: cmd/root.go [hunk 3]
```diff
@@ -28,35 +32,135 @@ type AIAssess struct {
...
+					fmt.Print("ollama not running on default port, attempting to start ollama...")
```
Comment (sean-hacker) [inline cmd/root.go]: Let's use the svc1log logger instead of Print

File: cmd/root.go [hunk 4]
```diff
@@ -28,35 +32,135 @@ type AIAssess struct {
...
+				errorMessage := err.Error()
+				a.OutputSignal.ErrorMessage = &errorMessage
+				a.OutputSignal.Status = 1
+				return fmt.Errorf(errorMessage)
+			}
+
+			if !ollama.ModelReady(ollamaURL, ollamaModel) {
+				if allowDownload {
+					// Download the model
+					err := ollama.DownloadOllamaModel(ollamaURL, ollamaModel)
```
Comment (sean-hacker) [inline cmd/root.go:130]: I don't _love_ that we allow arbitrary downloads. Maybe an allowlist of models?

File: cmd/root.go
Comment (sean-hacker) [inline cmd/root.go:1]: I added https://github.com/Method-Security/pkg/blob/develop/signal/signal.go#L45 to hopefully make life easier on all the error handling

File: internal/ollama/query.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+type ModelPromptContentGenerator func(string) string
+
+// ProcessContentRecursively processes the content recursively, splitting it if necessary.
+// The input always gets the same prompt generator call to ensure the instructions are consistent across splits.
+func ProcessContentRecursively(ctx context.Context, client *api.Client, model Model, input string, generator ModelPromptContentGenerator, combiner ModelPromptContentGenerator) (string, error) {
+	content := generator(input)
+	// Attempt to query the model
+	response, err := QueryModel(ctx, client, model, content)
+	if err != nil {
+		if IsContextLengthError(err) && len(input) > 100 {
```
Comment (sean-hacker) [inline internal/ollama/query.go]: why 100?

File: internal/ollama/setup.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+			if status, ok := result["status"].(string); ok {
+				progress <- status
+			}
+		}
+	}()
+
+	// Print progress updates
+	for status := range progress {
+		fmt.Println(status)
```
Comment (sean-hacker) [inline internal/ollama/setup.go]: Logger please

File: webassess
Comment (sean-hacker) [inline webassess:1]: Don't think you want to check this in. Also recommend adding it to gitignore.

File: internal/ollama/query.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+			// If context length is exceeded, split the content and process each half
+			mid := len(input) / 2
```
Comment (sean-hacker) [inline internal/ollama/query.go:63]: I think the naive version is ok for now, but you likely will want to do some HTML-aware processing try to chunk it at cleaner separators

File: internal/ollama/query.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+			// Recursively process the right half
+			rightResult, errRight := ProcessContentRecursively(ctx, client, model, rightInput, generator, combiner)
+			if errRight != nil {
+				return "", errRight
+			}
+
+			// Combine the results
+			combinedOutput := leftResult + "\n\n" + rightResult
+			combinedPrompt := combiner(combinedOutput)
```
Comment (sean-hacker) [inline internal/ollama/query.go]: Why would you not make the combiner a different function shape that takes in the two strings and does the combining itself? Feels like combiner should be of a different type anyways for code clarity

File: internal/ollama/query.go (hunk 7 above)
Comment (methodsecuritysj) [reply internal/ollama/query.go:63]: agreed

File: cmd/root.go
Comment (methodsecuritysj) [reply cmd/root.go:1]: is there an example of using it somewhere?

**Review by sean-hacker — DISMISSED** [review]

**Review by sean-hacker — APPROVED** [review]

**Pull request merged**